File: 56.R
# Powerful digit sum

# Constants
LIMIT <- 100

max_digit_sum <- 0

for ( a in seq.int(2, LIMIT) ) {
    digits <- rev(as.integer(strsplit(as.character(a), split=NULL)[[1]]))
    for ( b in seq.int(2, LIMIT) ) {
        new_digits <- vector()
        carry <- 0
        for ( i in seq.int(length(digits)) ) {
            prod <- digits[i] * a + carry
            last_digit <- as.integer(substr(prod, nchar(prod), nchar(prod)))
            new_digits <- c(new_digits, last_digit)
            if ( prod >= 10 ) {
                carry <- as.integer(substr(prod, 1, nchar(prod)-1))
            } else {
                carry <- 0
            }
        }
        if ( carry > 0 ) {
            carry_digits <- rev(as.integer(strsplit(as.character(carry),
                                                    split=NULL)[[1]]))
            new_digits <- c(new_digits, carry_digits)
        }
        digits <- new_digits
        if ( sum(digits) > max_digit_sum ) {
            max_digit_sum <- sum(digits)
        }
    }
}

cat(max_digit_sum, fill=TRUE)

File: 56.js
#!/usr/bin/env node

// Powerful digit sum

var LIMIT = 99;

var max_digit_sum = 0;
for (var a = 2; a <= LIMIT; a++) {
    var digits = a.toString().split('').reverse();
    for (var b = 2; b <= LIMIT; b++) {
        var new_digits = [];
        var carry = 0;
        for (var i = 0; i < digits.length; i++) {
            var product = parseInt(digits[i]) * a + (carry || 0);
            var last_digit_of_product = parseInt(product.toString().slice(-1));
            new_digits.push(last_digit_of_product);
            carry = parseInt(product.toString().slice(0, -1));
        }
        if (carry) {
            var carry_digits = carry.toString().split("").reverse();
            for (var i = 0; i < carry_digits.length; i++) {
                new_digits.push(parseInt(carry_digits[i]));
            }
        }
        digits = new_digits;
        var digit_sum = 0;
        for (var i = 0; i < digits.length; i++) {
            digit_sum += digits[i];
        }
        if (digit_sum > max_digit_sum) {
            max_digit_sum = digit_sum;
        }
    }
}

console.log(max_digit_sum);
